### PR TITLE
[V2 API] Parse arc into a Graph object

### PR DESCRIPTION
**Pull request by shifucun**

### CODE REVIEW

File: internal/server/v2/parser.go [hunk 1]
```diff
@@ -45,3 +47,91 @@ func splitArc(s string) ([]string, error) {
...
+	if len(s) == 0 {
+		return arc, nil
+	}
+	if s[0] == '[' {
+		if s[len(s)-1] != ']' {
+			return nil, status.Errorf(
+				codes.InvalidArgument, "invalid filter string: %s", s)
+		}
+		s = s[1 : len(s)-1]
+		// [prop1, prop2]
```
Comment (Spaceenter) [inline internal/server/v2/parser.go:76]: At this point, [ and ] are already removed, so:   prop1, prop2 ?

File: internal/server/v2/parser.go [hunk 2]
```diff
@@ -45,3 +47,91 @@ func splitArc(s string) ([]string, error) {
...
+		props := []string{}
+		pos := -1
+		for i := 0; i < len(s)-1; i++ {
+			if s[i] == ',' {
+				props = append(props, strings.TrimSpace(s[pos+1:i]))
+				pos = i
+			}
+		}
+		props = append(props, strings.TrimSpace(s[pos+1:]))
+		arc.props = props
```
Comment (Spaceenter) [inline internal/server/v2/parser.go:86]: A shorter version...

```
arc.props = strings.Split(strings.ReplaceAll(s, " ", ""), ",")
```

File: internal/server/v2/parser.go [hunk 3]
```diff
@@ -45,3 +47,91 @@ func splitArc(s string) ([]string, error) {
...
+		return arc, nil
+	}
+	for i := 0; i < len(s); i++ {
```
Comment (Spaceenter) [inline internal/server/v2/parser.go:89]: Add more comments on what each block is doing?

File: internal/server/v2/parser.go [hunk 4]
```diff
@@ -45,3 +47,91 @@ func splitArc(s string) ([]string, error) {
...
+		if s[i] == '+' {
+			arc.prop = s[0:i]
+			arc.wildcard = "+"
+			s = s[i+1:]
+			break
```
Comment (Spaceenter) [inline internal/server/v2/parser.go:94]: Is it possible to have the following pattern?

->prop1+{prop2: val2}

If so, probably 'break' is wrong here. If not, probably add a comment somewhere that the above pattern is not supported?

File: internal/server/v2/parser.go [hunk 5]
```diff
@@ -45,3 +47,91 @@ func splitArc(s string) ([]string, error) {
...
+		}
+	}
+	if len(s) > 0 && s[0] == '{' {
+		if s[len(s)-1] != '}' {
+			return nil, status.Errorf(
+				codes.InvalidArgument, "invalid filter string: %s", s)
+		}
+		// {prop1:val1, prop2:val2}
+		filter := map[string]string{}
+		parts := strings.Split(s[1:len(s)-1], ",")
```
Comment (Spaceenter) [inline internal/server/v2/parser.go:108]: parts := strings.Split(strings.ReplaceAll(s[1:len(s)-1], " ", ""), ",")

Then, you can remove all the strings.TrimSpace() calls below.

File: internal/server/v2/parser.go [hunk 6]
```diff
@@ -45,3 +47,91 @@ func splitArc(s string) ([]string, error) {
...
+			if len(kv) != 1 {
+				return nil, status.Errorf(
+					codes.InvalidArgument, "invalid filter string: %s", p)
+			}
+			k := strings.TrimSpace(p[0:kv[0]])
+			v := strings.TrimSpace(p[kv[0]+1:])
+			if len(k) == 0 || len(v) == 0 {
+				return nil, status.Errorf(
+					codes.InvalidArgument, "invalid filter string: %s", p)
+			}
```
Comment (Spaceenter) [inline internal/server/v2/parser.go:126]: Can we just use  strings.Split(p, ":") here? Then check if there are 2 non-empty items.

**Review by Spaceenter — CHANGES_REQUESTED** [review]

File: internal/server/v2/parser.go (hunk 1 above)
Comment (shifucun) [reply internal/server/v2/parser.go]: Moved the comment to the beginning of the block

File: internal/server/v2/parser.go (hunk 2 above)
Comment (shifucun) [reply internal/server/v2/parser.go]: much better!

File: internal/server/v2/parser.go (hunk 3 above)
Comment (shifucun) [reply internal/server/v2/parser.go:80]: Done

File: internal/server/v2/parser.go (hunk 4 above)
Comment (shifucun) [reply internal/server/v2/parser.go:86]: The above pattern should work. This block process upto "prop1+". Then from line101, it handles {prop2: val2}

File: internal/server/v2/parser.go (hunk 5 above)
Comment (shifucun) [reply internal/server/v2/parser.go]: updated!

File: internal/server/v2/parser.go (hunk 6 above)
Comment (shifucun) [reply internal/server/v2/parser.go]: that's definitely much better

File: internal/server/v2/graph.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+// Package v2 is the version 2 of the Data Commons REST API.
+package v2
+
+// Arc represents an arc in the graph.
+type Arc struct {
+	// Whether it's out or in arc.
+	out bool
+	// The property of the arc. This is when property is specified without []
+	prop string
```
[the remark this answers is not in the document]
Comment (shifucun) [reply internal/server/v2/graph.go]: Use "singleProp" and "bracketProps" to make the initial query intention clear. For "containedInPlace+" it uses single prop but implies a list of props, so it's better to have two terms to make this clear.

File: internal/server/v2/parser.go [hunk 8]
```diff
@@ -45,3 +47,73 @@ func splitArc(s string) ([]string, error) {
...
+	if len(s) > 0 && s[0] == '{' {
+		if s[len(s)-1] != '}' {
+			return nil, status.Errorf(
+				codes.InvalidArgument, "invalid filter string: %s", s)
+		}
+		filter := map[string]string{}
+		parts := strings.Split(strings.ReplaceAll(s[1:len(s)-1], " ", ""), ",")
+		for _, p := range parts {
+			kv := strings.Split(p, ":")
+			if len(kv) != 2 {
```
Comment (Spaceenter) [inline internal/server/v2/parser.go:105]: if len(kv) != 2 || kv[0] == "" || kv[1] == ""

**Review by Spaceenter — APPROVED** [review]

File: internal/server/v2/parser.go (hunk 8 above)
Comment (shifucun) [reply internal/server/v2/parser.go]: Done

**Pull request merged**